### PR TITLE
all: refactor handler logic

### LINKED ISSUE #306: authenticate: oauth csrf errors
Greetings and thanks for this nice project :)

I've installed Pomerium binary v0.3.0+bc72d08 on Debian 9.11

At first it was working, but after a few minutes I get only "Internal server errors", such as:
- oauth callback : csrf failure: %!w(<nil>)

I tried to restart Pomerium and clear cookies.

What did I do wrong?

Thanks in advance

### PR DESCRIPTION
**Pull request by desimone**
These changes move user authentication verification & handling, csrf protection, and service mutual authentication form handlers to middleware.

**Why**:

- CSRF mitigation (both between user sessions and from the identity provider) was difficult to reason about, sprawling, and error prone. 
- Session and csrf store were both implemented on top of cookie store in a confusing (and often overlapping) way. 
- Go's built in ServeMux has started to show it's limitations. In particular, #262 and lack of [named matched hosts](https://www.gorillatoolkit.org/pkg/mux#Route.Host) which required us to implement our own route hashmap. 
- User sessions were repeatedly being loaded from session 1) during middleware checks 2) during the handler itself. Now, subsequent per-request calls requiring user state can be fetched from context which is (imho) much more idiomatic.
- Enables the follow up (related) PRs coming for:
    - #282 
    - #262 
    - #281 
- Fixes #306

**Details**
- all: prefer `FormValues` to `ParseForm` with subsequent `Form.Get`s
- all: refactor authentication stack to be checked by middleware, and accessible via request context.
- all: replace http.ServeMux with gorilla/mux’s router
- all: replace custom CSRF checks with gorilla/csrf middleware
- authenticate: extract callback path as constant.
- internal/config: implement stringer interface for policy
- internal/cryptutil: add helper func `NewBase64Key`
- internal/cryptutil: rename `GenerateKey` to `NewKey`
- internal/cryptutil: rename `GenerateRandomString` to `NewRandomStringN`
- internal/middleware: removed alice in favor of gorilla/mux
- internal/sessions: remove unused `ValidateRedirectURI` and `ValidateClientSecret`
- internal/sessions: replace custom CSRF with gorilla/csrf fork that supports custom handler protection
- internal/urlutil: add `SignedRedirectURL` to create hmac'd URLs
- internal/urlutil: add `ValidateURL` helper to parse URL options
- internal/urlutil: add `GetAbsoluteURL` which takes a request and returns its absolute URL.
- proxy: remove holdover state verification checks; we no longer are setting sessions in any proxy routes so we don’t need them.
- proxy: replace un-named http.ServeMux with named domain routes.

**Checklist**:
- [x] updated docs
- [x] unit tests added
- [x] related issues referenced
- [x] updated CHANGELOG.md
- [x] ready for review

### CODE REVIEW

File: proxy/proxy.go [hunk 1]
```diff
@@ -34,29 +35,32 @@ const (
 	HeaderGroups = "x-pomerium-authenticated-user-groups"
 )
 
+var (
```
Comment (travisgroth) [inline proxy/proxy.go]: should these be const?

Comment (desimone) [reply proxy/proxy.go]: Yes, unfortunately Go doesn't support const structs (AFAIK). 

I split the difference and made the constant values strings and then cast them to URLs in the constructor body. Let me know what you think.

**Comment by desimone** [comment]
@travisgroth incorporated changes. let me know if looks good to go, if so I'll squash and merge.

File: proxy/proxy.go (hunk 1 above)
Comment (travisgroth) [reply proxy/proxy.go]: Ah.  Yeah that makes sense.

**Pull request merged**